keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -0,0 +1,165 @@
+2.  `LastPodGroup` records which `PodGroup` the last scheduled pod belongs to.
+3.  `UID` is the unique identification value used to distinguish different podgroups.
+ 
+
+### Extension points
+
+#### QueueSort
+In order to make the pods which belongs to the same `PodGroup` to be scheduled together as much as possible, implement a strategy in `QueueSort` phase. 
+
+**limition**: `QueueSort` is the core part of our design and only one `QueueSort` plugin is allowed in the scheduling framework. So our design only supports the case that `QueueSort` extension point isn't implemented in other plugins.
+
+```go
+  func  Less(podA *PodInfo, podB *PodInfo) bool
+```
+1. Trying to order by pod priority (i.e. .spec.priorityValue), pod with higher priority is scheduled ahead of other pod with lower priority. When priorities are the same, they are operated according to the following process. 
+   
+2. When podA and podB are both regularPods (we will check it by their labels), it follows the same logic of default in-tree [PrioritySort](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/queuesort/priority_sort.go#L41-L45) plugin.
+   
+3. When podA is regularPod, podB is pgPod, trying to order by podA's `Timestamp` (the time pod added to the scheduling queue) and podB’s `LastFailureTimestamp` (we get the PodGroupInfo from the cache). Pod with earlier timestamp is scheduled ahead of other pod.   
+   
+4. When podA and podB are both pgPods. 
+   1. Trying to order by the `LastFailureTimestamp` of the podGroup. Pod with earlier timestamp is scheduled ahead of other pod.
+   
+   2. If `LastFailureTimestampA` is equal to `LastFailureTimestampB`, trying to order by the `UID` of `PodGroup`,`Pod` with lexicographically greater `UID` is scheduled ahead of other pod. (The purpose is to distinguish different `PodGroup` with the same `LastFailureTimestamp` and to keep the pods of the same `PodGroup` together)
+
+**Note1**: There are different `LastFailureTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
+
+#### Pre-Filter
+1. `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.
+
+2. `PreFilter` validates that if we should reject `WaitingPods` in advance. When the pod belongs to a new `PodGroup` (we can check it by the `LastPodGroup` in cache), it indicates that the new PodGroup scheduling cycle has been entered. Then we will check if the last `PodGroup` meets the minAvailable, if not, we will reject it directly in advance. The purpose is 1.To avoid invalid waiting, 2.To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.
```

Comment:
Can you show me why rejecting on-going waiting PodGroup is mandatory? It seems that we should use a Set (instead of String) for on-going PodGroup as a new PodGroup's scheduling is independent on existing on-going one, isn't it?

If we keep rejecting existing on-going PodGroup, it can still lead to deadlock even if right now we put Permit at the end of scheduling cycle.

Reply:
> Can you show me why rejecting on-going waiting PodGroup is mandatory?

The purpose of this is to make the pods belonging to the same `PodGroup` adjacent to each other in the process, rather than being queued by other pods.  eg: podA1, podA2, podA3(PodGroupA min=3),  podB1, podB2, podB3(PodGroupB min=3)

1. PodA1 PodA2  failed (Insufficient resources in the current cluster)
2. PodA3 success to WaitingMap (Can't actually schedule successfully because min=3)
3. PodB1 PodB2 failed (Insufficient resources in the current cluster)
4. PodB3 success to WaitingMap
5. Due to timeout mechanism, PodA3 and PodB3 are in WaitingMap. The new scheduling order is PodA1 PodA2 PodB1 PodB2. So PodGroupA is queued by other pods.

The above logic is to reduce the number of scheduling failures. But If it will lead to deadlock because rejecting existing on-going PodGroup, I think it is not necessary. WDYT? @Huang-Wei

Reply:
Ah, I see the point, any Pod in a PodGroup cannot "reject" the on-going PodGroup - because you don't know if you're the last one and hence cannot tell whether reject or not. What a Pod in an on-going PodGroup can only "approve" - when the waitingPod number exceeds the minAvailable.

So the 2nd paragraph is more an optimization to fail fast the previous PodGroup - which is destined unable to be scheduled, please highlight that. And I think it's not necessarily only the (1st) Pod of another PodGroup can make this (reject) decision - any Pod other than the Pod belonging to the previous PodGroup can make the decision, right?

Reply:
So you are basically trying to hold a second PodGroup until the first PodGroup has satisfied minAvailable. I have the feeling that Wei has a different interpretation.

But then pods from the second group will move to the unschedulable queue. How do you plan to bring them back once the first group satisfies it's lower limit?

Reply:
> And I think it's not necessarily only the (1st) Pod of another PodGroup can make this (reject) decision - any Pod other than the Pod belonging to the previous PodGroup can make the decision, right?

Yes, that's right. It is just optimized mechanism to fail fast the previous PodGroup. But I think it is difficult to chose when to reject the previous PodGroup. I temporarily removed this mechanism from the KEP.

> So you are basically trying to hold a second PodGroup until the first PodGroup has satisfied minAvailable. 

Not to hold the second PodGroup until the first PodGroup has satisfied minAvailable. It is to fail the first PodGroup fast if it is determined the first PodGroup is unable to satisfy minAvailable competition because the second PodGroup is coming is schedule cycle.

Reply:
This is definitely not clear at all. Let's add the conditions for considering a pod group unsatisfiable.

But also I  don't think this is a good idea. Let's say some PG1 and PG2 (both requiring 3 min) are being created at almost the same time and the queue looks like this:

```
queue              | events about to come from apiserver
PG1, PG1, PG2, PG2 | PG1, PG2
```

The pods after the line are not in the queue yet, because there might be a delay from apiserver. But it seems that in your proposal, once the first PG2 is seen, we already reject PG1, because it's not satisfiable. But it will be soon.

Reply:
Yes,  you are right. There's something wrong with this mechanism, because it is difficult to chose when to reject the previous PodGroup. I remove this mechanism in `PreFilter`. Thank you @alculquicondor